types/valuerenderer/valuerenderer_test.go

```diff
@@ -0,0 +1,263 @@
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	"golang.org/x/text/language"
+	"golang.org/x/text/message"
+)
+
+const (
+	holder     = "holder"
+	multiPerm  = "multiple permissions account"
+	randomPerm = "random permission"
+)
+
+func initBankKeeperAndContext(t *testing.T) (keeper.BaseKeeper, types.Context) {
+	app := simapp.Setup(t, false)
+	c := app.BaseApp.NewContext(false, tmproto.Header{})
+	maccPerms := simapp.GetMaccPerms()
+	appCodec := simapp.MakeTestEncodingConfig().Codec
+
+	maccPerms[holder] = nil
+	maccPerms[authtypes.Burner] = []string{authtypes.Burner}
+	maccPerms[authtypes.Minter] = []string{authtypes.Minter}
+	maccPerms[multiPerm] = []string{authtypes.Burner, authtypes.Minter, authtypes.Staking}
+	maccPerms[randomPerm] = []string{"random"}
+	authKeeper := authkeeper.NewAccountKeeper(
+		appCodec, app.GetKey(banktypes.StoreKey), app.GetSubspace(banktypes.ModuleName),
+		authtypes.ProtoBaseAccount, maccPerms,
+	)
+	blockedAddrs := make(map[string]bool)
+
+	keeper := keeper.NewBaseKeeper(
+		appCodec, app.GetKey(banktypes.StoreKey), authKeeper,
+		app.GetSubspace(banktypes.ModuleName), blockedAddrs,
+	)
+
+	return keeper,c
+}
+
+// TODO add more test cases
+func TestFormatCoin(t *testing.T) {
+
+	bk, c := initBankKeeperAndContext(t)
+	ctx := types.WrapSDKContext(c)
+	dvr := valuerenderer.NewDefaultValueRenderer(bk)
+	p := message.NewPrinter(language.English)
+
+	// TODO add test case to convert from mregen to uregen
+	tt := []struct {
+		name   string
+		coin   types.Coin
+		metadata banktypes.Metadata
+		expErr bool
+	}{
+		{
+			"convert 1000000uregen to 1regen",
+			types.NewCoin("uregen", types.NewInt(int64(1000000))),
+			banktypes.Metadata{
+					Name:        "Regen",
+					Symbol:      "REGEN",
+					Description: "The native staking token of the Regen network.",
+					DenomUnits: []*banktypes.DenomUnit{
+						{
+							Denom:    "uregen",
+							Exponent: 0,
+							Aliases:  []string{"microregen"},
+						},
+						{
+							Denom:    "regen",
+							Exponent: 6,
+							Aliases:  []string{"REGEN"},
+						},
+					},
+					Base:    "uregen",
+					Display: "regen",
+			},
+			false,
+		},
+		/*
+		{
+			"convert 1000000000uregen to 1000regen",
+			types.NewCoin("uregen", types.NewInt(int64(1000000000))),
+			false,
+		},
+		{
+			"convert 23000000mregen to 23000regen",
+			types.NewCoin("mregen", types.NewInt(int64(23000000))),
+			false,
+		},
+		{
+			"convert 23000000mregen to 23000000000uregen",
+			types.NewCoin("mregen", types.NewInt(int64(23000000))),
+			false,
+		},
+		{
+			"convert 23000000regen to 23000000000mregen",
+			types.NewCoin("regen", types.NewInt(int64(23000000))),
+			false,
+		},
+		{
+			"convert 23000regen to 23000regen",
+			types.NewCoin("regen", types.NewInt(int64(23000))),
+			false,
+		},
+		*/
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			dvr.GetBankKeeper().SetDenomMetaData(c, tc.metadata)
```

Comment:
Just wondering if the logic is correct here

1. I set `metadata` to the `bankKeeper` in `SetDenomMetaData`.`Metadata` is persisted  at  `KVStore` under the hood.
2. I query `metadata` in `dvr.QueryDenomMetadata(ctx, tc.coin)` to get this `metadata` from `KVstore`
3. I set `metadata` into `DefaultValueRenderer` field in order to use it in `dvr.Format()` later at L129.
4. Then I check if  `res` matches concatanated `expAmount` and `expDenom`